### PR TITLE
[DRAFT] Initial SDCA regmap integration

### PR DESCRIPTION
**Pull request by charleskeepax**
Example of where we are getting to integrating the register map stuff and the SDCA stuff. This series pulls in most of @plbossart pull request #5128, some of his older changes that added more DisCo parsing, and my pull request #5087. As well as a brief update to the cs42l43 driver to let it use the new stuff. Its a lot of changes so probably better to look at individual patches rather than trying to make sense of the so useful mega diff Github steers you towards.

The code adds marking which registers are readable/writeable etc. from DisCo, adds the disco constants into the map so they can be read like any other register, and does some initial handling of DisCo defaults.

The two major changes in the SDCA side of things: 1) sdca_function_data is now a child of sdca_function_desc 2) parsing the disco is now handled before the auxiliary devices are probed. These two are basically in service of getting the whole regmap for the device functional before any of the auxiliary devices probe.

Still all a bit of a work in progress but this is a good point to have some discussions around direction and general stuff.

### CODE REVIEW

File: sound/soc/sdca/sdca_functions.c [hunk 1]
```diff
@@ -170,6 +171,84 @@ void sdca_lookup_functions(struct sdw_slave *slave)
...
+		if (!fwnode_property_read_u32(control_node,
+					      "mipi-sdca-control-access-mode",
+					      &clist))
+			entity->controls[j].mode = clist;
+
+		if (entity->controls[j].mode == SDCA_CONTROL_ACCESS_MODE_DC) {
+			if (!fwnode_property_read_u32(control_node,
+						      "mipi-sdca-control-dc-value",
+						      &clist))
+				entity->controls[j].value = clist;
```
Comment (plbossart) [inline sound/soc/sdca/sdca_functions.c:222]: global comment: there should probably need an error thrown if the required properties are not defined, found or there's some problem accessing it.
Fail big and fail early....

File: sound/soc/sdca/sdca_functions.c [hunk 2]
```diff
@@ -190,14 +269,14 @@ static int find_sdca_entities(struct device *dev,
 			__func__, "mipi-sdca-entity-id-list",
 			num_entities);
 		return -EINVAL;
-	}
-	if (num_entities >  SDCA_MAX_ENTITY_COUNT) {
+	} else if (num_entities >  SDCA_MAX_ENTITY_COUNT) {
```
Comment (plbossart) [inline sound/soc/sdca/sdca_functions.c:274]: we could fold this change in the initial patch, it's unrelated to the controls.

File: sound/soc/sdca/sdca_functions.c [hunk 3]
```diff
@@ -250,16 +329,30 @@ static int find_sdca_entities(struct device *dev,
 			}
 		}
 
-		fwnode_handle_put(entity_node);
-
```
Comment (plbossart) [inline sound/soc/sdca/sdca_functions.c]: nit-pick: we could move this after the dev_info in the initial patch to make the next change more obvious.

File: sound/soc/sdca/sdca_functions.c [hunk 4]
```diff
@@ -250,16 +329,30 @@ static int find_sdca_entities(struct device *dev,
...
+
 	}
 
-	function->num_entities = num_entities;
+	ret = find_sdca_entity_controls(dev, function_node, &entities[num_entities],
+					function->function_desc->type);
+	if (ret)
+		return ret;
+
+	entities[num_entities].label = "entity0";
```
Comment (plbossart) [inline sound/soc/sdca/sdca_functions.c:355]: nit-pick: one would have thought using the index0 for entity0 would be logical... Maybe we need a comment to say that the entities are not sorted by entity number.

File: sound/soc/sdca/sdca_function_device.c [hunk 5]
```diff
@@ -86,6 +86,33 @@ static void sdca_dev_unregister(struct sdca_dev *sdev)
...
+		func_desc->function = func;
+
+		ret = sdca_parse_function(dev, func_desc->function_node, func_desc);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+EXPORT_SYMBOL_NS(sdca_dev_parse_functions, SND_SOC_SDCA);
```
Comment (plbossart) [inline sound/soc/sdca/sdca_function_device.c:119]: this is a bit controversial. This will result in a layering violation, where the properties that are in scope of a Function device are now parsed at the parent (SoundWire peripheral) device level.

I am not sure why this is needed really. It's perfectly reasonable to have an SDCA regmap for each function device.

I guess if all the functionality is implemented at the SoundWire peripheral device level without creating any SDCA function subdevices then this approach would make sense.

This is a rather important architectural point that needs more pros/cons explanations.

File: sound/soc/sdca/sdca_function_device.c [hunk 6]
```diff
@@ -49,6 +51,7 @@ static struct sdca_dev *sdca_dev_register(struct device *parent,
 	auxdev->dev.fwnode = function_desc->function_node;
 	auxdev->dev.release = sdca_dev_release;
 
+	sdev->regmap = regmap;
```
Comment (plbossart) [inline sound/soc/sdca/sdca_function_device.c:56]: similar point, why would the parent device pass a regmap when a regmap can be created during the probe of an auxiliary device?

If the parent creates a giant regmap that's the union of all controls across all Functions, then it's a rather important risk factor where a control from one function might be modified in another.

The SDCA definitions guarantee that the addresses cannot overlap between functions.

We'd also need to clarify which regmap this is referring to. I was assuming this is the SDCA controls only, including the MBQ parts. I must admit not being sure of how to handle non-SDCA registers, e.g. if there is a vendor-specific register this could be fun.

File: include/sound/sdca_function.h [hunk 7]
```diff
@@ -54,6 +54,12 @@ enum sdca_entity0_controls {
 	SDCA_CONTROL_ENTITY_0_FUNCTION_VERSION		= 0x08
 };
 
+enum sdca_fu_controls {
+	SDCA_CONTROL_FU_MUTE				= 0x01,
+	SDCA_CONTROL_FU_CHANNEL_VOLUME			= 0x02,
+	SDCA_CONTROL_FU_LATENCY				= 0x10,
```
Comment (plbossart) [inline include/sound/sdca_function.h:61]: I wouldn't worry too much about latency, it was added in the spec because it was already in the USB spec, but no one seems to know how to account for latencies reported here. Best to ignore and focus on important features :-)

File: sound/soc/sdca/sdca_device.c [hunk 8]
```diff
@@ -60,3 +62,185 @@ bool sdca_device_quirk_match(struct sdw_slave *slave, enum sdca_quirk quirk)
...
+	entity = sdca_function_find_entity(function, reg);
+	if (!entity)
+		return NULL;
+
+	return sdca_entity_find_control(entity, reg);
+}
+
+static bool sdca_valid_address(unsigned int reg)
+{
+	if ((reg & (GENMASK(31, 25) | BIT(18) | BIT(13))) != BIT(30))
```
Comment (plbossart) [inline sound/soc/sdca/sdca_device.c:124]: It wouldn't hurt to add a comment for humans.

File: sound/soc/sdca/sdca_function_hid.c [hunk 9]
```diff
@@ -17,6 +17,14 @@
...
 			  const struct auxiliary_device_id *aux_dev_id)
 {
+	struct sdca_dev *sdev = auxiliary_dev_to_sdca_dev(auxdev);
+	struct device *dev = &auxdev->dev;
+	int ret;
+
+	ret = sdca_disco_write_defaults(sdev);
+	if (ret)
+		return dev_err_probe(dev, ret, "Failed to write default values\n");
+
```
Comment (plbossart) [inline sound/soc/sdca/sdca_function_hid.c:27]: Now this is odd. The parsing was done at the SoundWire device level, and the regmap is passed to the aux device during registration, but the management of default values is now handled during the SDCA aux device probe..

Humm, that doesn't seem right.

File: drivers/pinctrl/cirrus/Kconfig [hunk 10]
```diff
@@ -1,15 +1,4 @@
...
-	tristate "Cirrus Logic CS42L43 Pinctrl Driver"
-	depends on MFD_CS42L43
-	select GPIOLIB
-	select PINMUX
-	select PINCONF
-	select GENERIC_PINCONF
-	help
-	  Select this to support the GPIO/Pinctrl functions of the Cirrus
-	  Logic CS42L43 PC CODEC.
-
```
Comment (plbossart) [inline drivers/pinctrl/cirrus/Kconfig:12]: You probably need a different way of handling the changes that roll-back and re-add. This would be problematic for any git bisect with intermediate commits removing functionality.

File: drivers/mfd/cs42l43.c [hunk 11]
```diff
@@ -805,6 +829,26 @@ static void cs42l43_boot_work(struct work_struct *work)
...
+					      &cs42l43_mbq_config);
+	if (IS_ERR(regmap))
+		return;
+
+	ret = sdca_dev_register_functions(cs42l43->sdw, regmap);
+	if (ret) {
+		dev_err(cs42l43->dev, "Failed to add subdevices: %d\n", ret);
+		goto err;
+	}
+
```
Comment (plbossart) [inline drivers/mfd/cs42l43.c:851]: Again I am not sure why a single regmap across all functions would make sense, compared to a regmap allocated during the probe of each SDCA function driver.

I don't see anything done here that couldn't be done at a lower level.

**Review by plbossart — COMMENTED** [review]
Lots of good idea and useful helpers @charleskeepax 

The one big open is the notion that there would be a single SDCA regmap shared across all functions. That seems scary and I don't see any benefit. I must be missing something important that drove you to choose that design?

File: sound/soc/sdca/sdca_functions.c (hunk 2 above)
Comment (charleskeepax) [reply sound/soc/sdca/sdca_functions.c:274]: Yeah certainly makes sense was mostly trying to minimise the number of changes I folded into your patch whilst they are still separate chains. But yeah would make more sense folded into your one.

File: sound/soc/sdca/sdca_functions.c (hunk 1 above)
Comment (charleskeepax) [reply sound/soc/sdca/sdca_functions.c:222]: Yeah mostly this is still fairly early stage stuff for a lot of things we should return a few more errors etc., although some of it might require some thinking as I have a feeling errors might not be that rare and we generally don't get to fix the ACPI, but at least some bit fat warnings would be nice.

File: sound/soc/sdca/sdca_functions.c (hunk 4 above)
Comment (charleskeepax) [reply sound/soc/sdca/sdca_functions.c:355]: Yeah I basically went this way for two reasons, 1) it makes the code slightly easier to follow, 2) entity0 is a bit of a fake entity doesn't have like connections etc. so seems mild more efficient if they entities you are more likely to want to access are earlier in the array. Regarding the sorting its populated from the ACPI anyway so there is no guarantee it is sorted anyway, no objections to adding a comment to that effect though.

File: drivers/pinctrl/cirrus/Kconfig (hunk 10 above)
Comment (charleskeepax) [reply drivers/pinctrl/cirrus/Kconfig:12]: There is no intention of actually updating cs42l43 to use the SDCA stuff certainly in the short term and even if we did we would likely do so in a keeping both options available way. This is just to provide me a test platform.

File: include/sound/sdca_function.h (hunk 7 above)
Comment (charleskeepax) [reply include/sound/sdca_function.h:61]: Yeah obviously like a million more controls to add to this in the long term, this was just a little set that handily is enough to get my smart amp to produce sound and gives me a register of each MBQ size.

File: sound/soc/sdca/sdca_function_device.c (hunk 5 above)
Comment (charleskeepax) [reply sound/soc/sdca/sdca_function_device.c:119]: Yeah again commit messages are still very lacking in this chain, apologies for that. I thought this might be the primary point of discussion between us on this stuff.

Firstly to be clear my intention is very much to have subdevices for each function, just to have them all share a common register map.

There are few reasons this makes sense, primarily its a single hardware device and generally one wants a register map to correspond to the hardware device. Only a single thing can talk to the hardware at one time, so it should be locked as such. Especially so in this case where SDCA only mandates a single MBQ buffer for the whole device, rather than one per function, which means you need several actual bus transactions to be atomic and not interrupted by another function. If you were to add a register map for each function you would have to have a lock at the soundwire device level that each subdevice took before it accessed the register map which is rather messy and error prone.

Secondly there are situations where one might want to access registers, especially the function level registers, before registering the subdevices. Currently the way we have this laid out, as per my comment on your patch chain, we force the function-type control to be a disco constant, however it would be totally spec compliant for that to be a device register. Creating the register map at this level would allow us to access these values as registers regardless of if they are a device register or a disco constant. Although I still need to work through your quirks issues and see if we can avoid forcing the constants there.

In short I see a lot of problems with creating a register map for each function and the only really advantage one gets is that it is slightly harder for one function to accidentally access another functions registers, which a function shouldn't be doing anyway. If a function is accessing the wrong registers it is likely the system is already not working properly.

File: sound/soc/sdca/sdca_function_hid.c (hunk 9 above)
Comment (charleskeepax) [reply sound/soc/sdca/sdca_function_hid.c:27]: Its important to separate what the two are doing, the soundwire device level operation creates the register map and parses any constant values adding them to the register map such that devices can read them. That operation doesn't actually cause any interaction with this device at all. This level actually writes out values to the hardware. Whilst I would agree there is some debate about which level to place this, it seemed to me there was no point in writing out configuration to functions that are not getting probed, unless anyone has a good reason we should?

But basically the partition I was looking at is we initialise the register map at the soundwire level, but then accesses to specific functions should all come from that function.

File: drivers/pinctrl/cirrus/Kconfig (hunk 10 above)
Comment (charleskeepax) [reply drivers/pinctrl/cirrus/Kconfig:12]: Worth saying I included the code in the chain in case you guys wanted to try things out, this should all work with the cs42l43's you have. Although you would need some more changes I haven't pushed yet to get audio out, but this will probe the device and setup the register map.

File: sound/soc/sdca/sdca_function_device.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sdca/sdca_function_device.c:119]: The premise has always been that Functions are independent.

The fact that there are choke points at the device level such as interrupt handling and MBQ buffers should not be a reason to rejoin functions together.

The mutual exclusion for MBQ access is not hard to add, it's one lock at the device level that needs to be taken/released in regmap_sdw_mbq_write/read.  It's simpler IMHO than creating a giant regmap that includes all the SDCA controls, even if/when there's no function driver probed.

That said, if for some reason a function driver needs to access a vendor-specific register that's not in the SDCA space, then I have no idea what to do. Even for SDCA memories for e.g. firmware download the regmap directions are not clear to me.

Comment (plbossart) [reply sound/soc/sdca/sdca_function_device.c:119]: Re: the function-level registers that would need to be accessed, if they are read-only they might as well be DC constant. If they are read-write, then *something* needs to have write access before the OS boots, and that adds a whole new level of complexity that is probably unmanageable with DisCo. You cannot discover information from ACPI and from the hardware, this doesn't work. Either all the information is read from hardware (USB, HDaudio) or we rely on ACPI. So if we are going to rely on DisCo, I think we can safely assume that all resources needed for driver probe are DC constant and defined by ACPI.

Comment (charleskeepax) [reply sound/soc/sdca/sdca_function_device.c:119]: We are not joining functions together just adding a common mechanism for accessing the hardware. Sharing the register map gets you all those things for free. Passing extra locks down into the register map is much worse layering than having a single mechanism to access the hardware. What problems are we actually trying to avoid with the separate register maps? All it does is add complexity and opportunities for disaster.

Re-the function level registers, they can be read only, but not disco constant, you can have variations on the hardware.

Comment (plbossart) [reply sound/soc/sdca/sdca_function_device.c:119]: well in hindsight that's another layering violation. We should have implemented the multiple MBQ read/writes in the SoundWire core with the proper API and locking. We currently did all this at the regmap layer by piggy-backing on regmap locking, so now regmap is the answer to everything. When all you have is a hammer, etc.

I really don't think it would be a 'disaster' to keep registers in the scope in which they are defined, and to deal with the transport separately.

Another practical aspect is that it'll be rather unlikely that all functions in a multi-function device are used at the same time. So if you only want to use the mic function, it'd be handy to only have to deal with regmap for the mic part only. Do we really want to do a regmap_sync() for everything or just for what we need when resuming?

Comment (charleskeepax) [reply sound/soc/sdca/sdca_function_device.c:119]: Hm... will have a think about what MBQing at the SoundWire level would look like, although the MBQ stuff has always felt like a protocol that sits on top of SoundWire to me, which is why I chose the regmap layer to put it in. If anything since we added the function_status stuff into it rather than the retrying I was nervous it was a bit low in the hierarchy, since we are now assuming the function_status control exists, which granted it is mandatory but you know people do things. Directly assuming SDCA controls at the soundwire level feels like we are poking up through the layers quite a lot.

Syncing the registers is a fair point in favour of multiple register maps, it would be nice to sync less, although in most practical situations I am not sure we are talking about that many registers here but it is probably worth investigating.

Comment (plbossart) [reply sound/soc/sdca/sdca_function_device.c:119]: There's also the case where the Function may set the Function_Needs_Initialization bit. In those cases it'd make sense to first update regmap then do a regmap_sync().

Note that the aux device model is missing something for the initial boot. The regmap creation, be it at the device or function level, takes place during a probe. When the SoundWire device actually shows up on the bus, there's a need to exit the regmap-cached mode. If the regmap handling is done in the function driver, then we're missing a signal sent to all subdevices.

Comment (charleskeepax) [reply sound/soc/sdca/sdca_function_device.c:119]: At the risk of starting another argument :-) I would argue it makes sense to only register the subdevices once the SoundWire device has shown up on the bus. This is what the cs42l43 example there does. This has many advantages meaning a lot of the hacks around waiting for the first time the device shows up on the bus (that we have in a lot of the current SoundWire devices) are no longer necessary. It also means you don't get madness like the soundcard being available to user-space but it not actually being able to do anything.

The init writes are also an interesting point, they feel very much like they should be modeled as a register_patch. The primary reason being they are likely to access the registers in a weird direct way (ie. either open coding the MBQ or not using the MBQ at all and directly addressing each byte). So it almost certainly makes sense to bypass the cache with those and just have them go straight to the hardware, as in a register patch.

It does raise an interesting question in my head of: can a function that has init writes, not require initialisation but still have lost the register values? ie. implying you want to sync but not initialise.

Comment (plbossart) [reply sound/soc/sdca/sdca_function_device.c:119]: There are pros and cons on the subdevice registration in the probe or update_status callback

probe:
- pros: all error handling is easier to do, fail big and early(tm). 
- cons: need to deal with the device being attached on the bus

update_status
- pros: no need to deal with additional signaling
-  cons: error handling is really messy.

The main reason why we used the regmap cache mode is so that we could use devm_ in the peripheral probe, that wouldn't work in the update_status callback.

But if you register a device, then devm_ is allowed in the probe, so it's only the error handling that's messy.

the init writes cannot use the MBQ, the format is a list of pairs of 32-bit address and 8 bit data. I guess you could encode the MBQ sequence but that would have to be split in N transactions. Deferred writes are not allowed either.

In theory according to the spec the Function_Needs_Initialization bit might be set at any time, but it's hard to think of a design where this would be required.

**Comment by lgirdwood** [comment]
@charleskeepax any update here ? I want to add in @bardliao and @ujfalusi too so we can all align now that Pierre is away.

**Comment by charleskeepax** [comment]
Not much update really, been a bit busy with support stuff of late, hopefully should get to start looking at this again next week. Will close this pull request anyway, as it will need a rebase on top of the stuff Bard merged the other day for the beginnings of the SDCA support.
